### DISCUSSION TITLE
LLM-powered `thefuck` in fish-shell?

### DISCUSSION THREAD

**Discussion by ifsheldon** [fish-shell/fish-shell#10976] [Ideas]
Hi! 

I'd really like to see a [`thefuck`](https://github.com/nvbn/thefuck) powered by a large language model, solving commands that somehow went wrong.

I'd love to contribute on this feature. Thanks to your effort of porting fish-shell to Rust, I can comfortably contribute to this project. 

I have a few questions, though. 

How can I get access to the output history of a command? For example, suppose our LLM-powered `thefuck` is named `heck`
```shell
$ git puhs

git: 'puhs' is not a git command......

$ heck

LLM: you seem to have made a typo, you can run `git push` instead.

run 'git push' (y/N): 
```
How can `heck` get the output of the previous `git` command?

[`thefuck`](https://github.com/nvbn/thefuck) reruns the previous command in a subprocess except for its experimental instant model, see the [README](https://github.com/nvbn/thefuck?tab=readme-ov-file#experimental-instant-mode).

I think we can just save the output of the last executed command somewhere and get the content via API. Of course, there will be privacy concerns, so I think it will be a functionality to be opt in and enabled. This privacy concern is also the reason why I'd like this feature to be built into fish-shell, using fish-shell's private APIs, instead of developing a 3rd party extension.

**Comment by faho** [comment]
No, thanks.

I trust LLMs as far as I can throw them and want them nowhere near fish.

(tbh personally I don't like that form of typo correction to begin with, but putting it into the lie machine is even worse)

As an aside:

>  I think it will be a functionality to be opt in and enabled

We try not to make cool things opt-in. If it needs to be opt-in because of "privacy concerns", then it just shouldn't be part of the shell.

I'm sorry, "privacy concerns" don't belong anywhere near a shell.

> I'd like this feature to be built into fish-shell, using fish-shell's private APIs, instead of developing a 3rd party extension

Having it in fish would not in any way assuage any privacy concerns.

**Reply by faho** [reply]
If you wish to build it as a 3rd party thing, hook it into [event handlers](https://fishshell.com/docs/current/language.html#event-handlers) - fish_postexec and fish_posterror should be of interest there.

> How can heck get the output of the previous git command?

The *output* of the previous git command never reaches fish. Output does not go through the shell.

The shell connects the terminal directly to the command.
